### PR TITLE
Chronos: add xshards_tsdataset.impute and test code

### PR DESCRIPTION
**Pull request by smurf-1119**

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/data/experimental/xshards_tsdataset.py [hunk 1]
```diff
@@ -269,6 +270,29 @@ def roll(self,
...
+               "linear": impute by linear interpolation.
+        :param const_num: only effective when mode is set to "const".
+
+        :return: the xshardtsdataset instance.
+        '''
+        
+        self.shards = self.shards.transform_shard(impute_timeseries_dataframe,
+                                                        dt_col, mode,
+                                                        const_num)
+       #  self.shards.reset_index(drop=True, inplace=True)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/experimental/xshards_tsdataset.py:293]: You can't call `reset_index` directly on `self.shards`. Here is a way you might be able to do it.

```python
def df_reset_index(df):
    df.reset_index()
    return df
self.shards.transform_shard(df_reset_index)
```

File: python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py [hunk 2]
```diff
@@ -192,6 +192,33 @@ def test_xshardstsdataset_roll_multiple_id(self):
...
+                                               extra_feature_col=["extra feature"], id_col="id")
+
+        with pytest.raises(RuntimeError):
+            tsdata.to_xshards()
+
+        # impute train
+        tsdata.impute("datetime", 'last')
+        horizon = random.randint(1, 10)
+        lookback = random.randint(1, 20)
+        tsdata.roll(lookback=lookback, horizon=horizon)
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py:209]: I think you don't need to roll here. You may just collect the df to driver and check if the df is correctly imputed.

File: python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py [hunk 3]
```diff
@@ -192,6 +192,33 @@ def test_xshardstsdataset_roll_multiple_id(self):
         x = np.concatenate([collected_numpy[i]['x'] for i in range(len(collected_numpy))], axis=0)
         assert x.shape == ((50-lookback-horizon+1)*2, lookback, 2)
 
+    def test_xshardstsdataset_impute_multiple_id(self):
+        shards_multiple = read_csv(os.path.join(self.resource_path, "multiple.csv"))
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py:196]: maybe generate a df that needs to be imputed(have missing values). dump it to tmp dir and load it back through `read_csv`. I think "multiple.csv" does not have any missing values.

File: python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py [hunk 4]
```diff
@@ -192,6 +192,20 @@ def test_xshardstsdataset_roll_multiple_id(self):
         x = np.concatenate([collected_numpy[i]['x'] for i in range(len(collected_numpy))], axis=0)
         assert x.shape == ((50-lookback-horizon+1)*2, lookback, 2)
 
+    def test_xshardstsdataset_impute(self):
+        for val in ["last", "const", "linear"]:
+            shards_tmp = read_csv(os.path.join(self.resource_path, "impute_test.csv"))
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/data/experimental/test_xshardstsdataset.py:197]: I think we can generate the csv and dump the csv only for temp use. We don't need to upload the `impute_test.csv`

File: python/chronos/src/bigdl/chronos/data/experimental/xshards_tsdataset.py [hunk 5]
```diff
@@ -269,6 +270,37 @@ def roll(self,
...
+        '''
+        
+        def df_reset_index(df):
+              df.reset_index(drop=True, inplace=True)
+              return df
+       
+        self.shards = self.shards.transform_shard(impute_timeseries_dataframe,
+                                                        self.dt_col, mode,
+                                                        const_num)
+        self.shards.transform_shard(df_reset_index)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/experimental/xshards_tsdataset.py:301]: if we need to `self.shards = self.shards.transform_shard(df_reset_index)`? I am not quite sure, please check.

**Review by TheaperDeng — APPROVED** [review]
LGTM, please run jenkins

**Comment by smurf-1119** [comment]
http://10.112.231.51:18888/view/BigDL-2.0-PR/job/BigDL-Chronos-PR-Validation/992/

**Pull request merged**